src/compiler.rs

```diff
@@ -75,14 +75,39 @@ impl Compiler {
               .join(Self::expand_tilde(&relative.cooked)?)
               .lexiclean();
 
-            if import.is_file() {
+            let import_base_str = import.to_string_lossy();
```

Comment:
`just` is expected to work with non-unicode paths, so we can't do a lossy conversation here. I'd suggest checking out the `globset` crate, which I believe handles non-unicode paths.

Reply:
`globset` unfortunately doesn't have a method to iterate over a file system entry and select only paths that match the glob, like the iterator that `glob_with` returns with the `glob` crate. Looking at how `glob_with` is implemented in `glob`, it seems like this code is a bit tricky to write and I'd rather avoid it (or if I did write that code, I think it might work out to just implementing path-globbing functionality from scratch, which is also undesirable). 

There's also this crate: https://crates.io/crates/wax , which is utf-8 only but claims to reasonably handle sane non-utf-8 text internally, and also seems to have the file system iterator I want. It also hasn't been updated in a bit and claims to be experimental, so I'm kind of iffy about this crate.

Maybe I'm missing something obvious that would make this easier to do.

Reply:
Ahh, that's lame. A concern about `wax`, any syntax that `wax` supports becomes part of the public-facing interface of `just`, and so can't be added and removed without a backwards-incompatible change. `wax` supports some pretty exotic syntax, like repetitions `<a*/:0,>` and  alternatives `{a?c,x?z,foo}`, which I doubt would be incredibly useful for use in import paths.

Reply:
Actually it looks like even globset expects to be given the glob as a string: https://docs.rs/globset/latest/globset/struct.GlobBuilder.html#method.new . Also see this 2019 comment from the maintainer of `globset` on `glob`: https://github.com/rust-lang/glob/issues/78#issuecomment-483733581 and this `globset` issue: https://github.com/BurntSushi/ripgrep/issues/1250 . 

I'm starting to think that there's no principled way to support globbing on non-utf8 filesystems without basically submitting a patch to some upstream glob project to do it, but maybe few enough people are using these tools with non-utf8 globs that no one has previously run into it?

The `to_string_lossy` function turns any non-utf8 sequence into the unicode replacement character �. I wonder if there is some other function that could be called to try to convert a `PathBuf` into a meaningful Rust utf-8 string (maybe doing something smart with guessing the original encoding). The quick solution would be to move this conversion into the `has_glob` block, which would mean that people using glob imports and file paths with non-utf8 names would run into issues but at least no existing just behavior would break.

Reply:
It's a bit lame, but we could just throw an error if you try to glob with a path which can't be converted to unicode. Non-unicode paths are very rare and basically just crazy, so not supporting them is probably fine, especially if they're only unsupported if you use a pretty niche feature.

I'm tempted to land this feature as unstable initially, both because of the unicode path issue, but also because there's some choice about which glob syntaxes we support, and it might be good to make compare them all and make a principled choice about what to support.